### ISSUE TITLE
VS Studio 2017 can build sln but CLI needs a workaround to build

### ISSUE THREAD

**Issue by wiiNinja** [dotnet/cli#7115]
## Steps to reproduce

I have a solution that contains 25 projects. If I use the command line to build:

dotnet clean mysolution.sln
dotnet restore mysolution.sln
dotnet build mysolution.sln

The build will fail for just one particular project. The first error I see from the build is:

C:\Program Files\dotnet\sdk\1.0.4\Sdks\Microsoft.NET.Sdk\build\Microsoft.PackageDependencyResolution.targets(154,5): error : Assets file 'D:\mtp\tmp2\server\myproj1\obj\project.assets.json' not found. Run a NuGet package restore to generate this file. [D:\mtp\tmp2\server\myproj1\myproj1.csproj]

Then there would be a bunch of similar errors like these:

C:\Temp\.NETCoreApp,Version=v1.1.AssemblyAttributes.cs(4,20): error CS0400: The type or namespace name 'System' could not be found in the global namespace (are you missing an assembly reference?) [D:\mtp\tmp2\server\myproj1\myproj1.csproj]

If I build using VS 2017, everything builds fine. If I build using VS 2017 at least once, subsequently, the CLI will also build successfully.

The workaround: If I build myproj1.csproj

cd myproj1
dotnet restore
dotnet build

Then the solution will also build fine.

## Expected  behavior

Both VS2017 and CLI to produce the same successful build.

## Actual behavior

The CLI failed to build the solution because for some reason it cannot build one particular project. Workaround by manually build that project before building the solution.

## Environment data
`dotnet --info` output:

.NET Command Line Tools (1.0.4)

Product Information:
 Version:            1.0.4
 Commit SHA-1 hash:  af1e6684fd

Runtime Environment:
 OS Name:     Windows
 OS Version:  6.1.7601
 OS Platform: Windows
 RID:         win7-x64
 Base Path:   C:\Program Files\dotnet\sdk\1.0.4

**Comment by livarcocc** [comment]
Can you share a small repro repo of this issue? What kinds of projects do you have in your solution? Are they all SDK based projects: projects with an SDK attribute in the Project Node?
The dotnet CLI only supports building projects that are SDK projects. If you have a mix of SDK and old style projects, you need to build with msbuild (which is what VS does). I wonder if that's why this is succeeding for you.

**Comment by livarcocc** [comment]
Given that I have heard back with more details on this in a few days, I am closing this issue. Please, re-activate when you can provide the details I asked above.

**Comment by wiiNinja** [comment]
Sorry I didn't get a chance to reply to this sooner, but because this is proprietary code, I do not have an option to post it. So I had to spend time experimenting to find a workaround. Thank you for looking into this issue.
So our solution consists of just SDK projects, however, I was told that the original solution file was generated from a previous project that "may" have the old style project (as you put it). The workaround I found was to use CLI to remove the offending projects and then add them again. This allows the build to go through properly. Looking at the sln file before and after the mentioned workaround, I see that offending projects has this section:

ProjectSection(ProjectDependencies) = postProject
	{F56F0E70-1306-4F56-8DC4-57B1FFDFD77E} = {F56F0E70-1306-4F56-8DC4-57B1FFDFD77E}
	{908B74FD-7658-4F83-BDFD-C1F94E3B91BB} = {908B74FD-7658-4F83-BDFD-C1F94E3B91BB}
EndProjectSection

while the non-offending projects do not. I looked here for some reference:

https://docs.microsoft.com/en-us/visualstudio/extensibility/internals/solution-dot-sln-file

but could not find any details on the "ProjectDependencies". Do you know how this may contribute to the problem I'm seeing?
Thank you.

**Comment by livarcocc** [comment]
@rainersigwald would you know? I am not super familiar with sln files format.

**Comment by rainersigwald** [comment]
MSBuild does have special code to handle solution dependencies, but I don't know of any problems that would result in this behavior.

I just tried setting up a simple solution with SDK projects and solution build dependencies, and it seemed to restore and build fine, so that's probably not (all of) the thing that's causing your problem.

@wiiNinja Do you notice the project type changing after remove + re-add? It's the first GUID listed in lines like this:

```sln
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ClassLibrary2", "ClassLibrary2\ClassLibrary2.csproj", "{48C9AE25-A853-493F-89C8-C0ED290921B4}"
```

Or, alternately, is there a change in the configuration section:
```sln
        GlobalSection(ProjectConfigurationPlatforms) = postSolution
                {F68E9500-BDFD-4782-84C3-E27690C38825}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
                {F68E9500-BDFD-4782-84C3-E27690C38825}.Debug|Any CPU.Build.0 = Debug|Any CPU
                {F68E9500-BDFD-4782-84C3-E27690C38825}.Release|Any CPU.ActiveCfg = Release|Any CPU
                {F68E9500-BDFD-4782-84C3-E27690C38825}.Release|Any CPU.Build.0 = Release|Any CPU
                {16EF7556-84ED-4185-8FD1-298975670203}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
                {16EF7556-84ED-4185-8FD1-298975670203}.Debug|Any CPU.Build.0 = Debug|Any CPU
                {16EF7556-84ED-4185-8FD1-298975670203}.Release|Any CPU.ActiveCfg = Release|Any CPU
                {16EF7556-84ED-4185-8FD1-298975670203}.Release|Any CPU.Build.0 = Release|Any CPU
                {48C9AE25-A853-493F-89C8-C0ED290921B4}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
                {48C9AE25-A853-493F-89C8-C0ED290921B4}.Debug|Any CPU.Build.0 = Debug|Any CPU
                {48C9AE25-A853-493F-89C8-C0ED290921B4}.Release|Any CPU.ActiveCfg = Release|Any CPU
                {48C9AE25-A853-493F-89C8-C0ED290921B4}.Release|Any CPU.Build.0 = Release|Any CPU
        EndGlobalSection
```

? Perhaps the projects were set up in the default configuration in such a way that NuGet's solution restore passed over them.

It might be helpful if you'd be willing to share just your .sln files (ideally both "before" and "after" versions). That will, of course, have all of your project names and possibly some other sensitive information in it. If you don't want to share it in public, you can email me at my microsoft.com address in my profile. If you don't want to do that either, I totally understand but we might not be able to get to the bottom of this.

**Comment by wiiNinja** [comment]
@rainersigwald - Thanks for looking into this.
I did see a change in the GUIDs for the removed/re-added projects, and assumed that the GUID is normally uniquely created when a project is added to a solution file. Are you saying that the GUID is unique to that project and should not change?
A few additional observations:
1) I can use VS2017 to remove, then add the projects, and the mentioned "ProjectDependencies" section will be wiped, the same as what the CLI option produces.
2) With the offending projects, I can manually remove the ProjectDependencies sections, and the solution will build correctly (don't have to use VS or CLI)
3) I have also created solution files from scratch using simple dummy projects, but could never reproduce this issue. I'm just wondering if these ProjectDepencies sections were residual from an earlier creation of the sln file that need not be there, and potentially can cause problem.
4) I can successfully build the offending solution file using Dotnet 2.0 Preview2.
5) As an experiment, I added a new dummy project to the solution using VS and attempted to add dependencies to the project, but I could not get the solution file to contain those ProjectDepencies lines, which leads me to believe that those are not needed or obsoleted somehow.

Regarding sending the before/after solutions files. I have prepared a set of sln files, but I need to check with a couple of people here before sending/posting. If not the whole solution files, perhaps I can post the relevant portions that have changed.
